### PR TITLE
Reschedule with soft image affinity 

### PR DESCRIPTION
**Pull request by jimmyxian**
Implement : https://github.com/docker/swarm/issues/743

>1/ Try a create. It fails with 404
>2/ Try a pull. It fails with something like image not found
>3/ Check if the image exists in the cluster.
>4/ Retry with a soft-affinity (but don't store it in the ContainerConfig)
>5/ If it still fails, then there is a problem with the image, abort
>6/ If it works, celebrate

Signed-off-by: Xian Chaobo <xianchaobo@huawei.com>

### CODE REVIEW

**Comment by jimmyxian** [comment]
ping @aluzzardi @vieux

File: cluster/config.go [hunk 1]
```diff
@@ -146,3 +146,18 @@ func (c *ContainerConfig) Affinities() []string {
 func (c *ContainerConfig) Constraints() []string {
 	return c.extractExprs("constraints")
 }
+
+// AddSoftImageAffinity to config
+func (c *ContainerConfig) AddSoftImageAffinity(image string) error {
+	affnitiesJSON := c.Labels[SwarmLabelNamespace+".affinities"]
```
Comment (vieux) [inline cluster/config.go]: you can use `c.extractExprs("affinities")`

File: cluster/config.go [hunk 2]
```diff
@@ -146,3 +146,18 @@ func (c *ContainerConfig) Affinities() []string {
 func (c *ContainerConfig) Constraints() []string {
 	return c.extractExprs("constraints")
 }
+
+// AddSoftImageAffinity to config
+func (c *ContainerConfig) AddSoftImageAffinity(image string) error {
```
Comment (aluzzardi) [inline cluster/config.go]: I would call this `AddAffinity` and pass `image==~<image>` from the caller.

File: cluster/config.go (hunk 1 above)
Comment (jimmyxian) [inline cluster/config.go]: OK

File: cluster/config.go (hunk 2 above)
Comment (jimmyxian) [inline cluster/config.go]: Yeah..I will fix it

**Comment by jimmyxian** [comment]
ping @aluzzardi @vieux 
Review again, thanks

**Comment by vieux** [comment]
LGTM

**Comment by aluzzardi** [comment]
Amazing :)

LGTM

We probably need to do some manual testing, just in case.

/cc @abronan

**Comment by abronan** [comment]
@vieux @aluzzardi Tested on my cluster, work like a charm :+1: 

LGTM

**Comment by abronan** [comment]
Thanks @jimmyxian! :)

**Pull request merged**